### ISSUE TITLE
ENOENT once.js on yarn install

### ISSUE THREAD

**Issue by code0100fun** [yarnpkg/yarn#2915]
**Do you want to request a *feature* or report a *bug*?**
A bug

**What is the current behavior?**
A fresh global install of yarn fails to `yarn install` in a project due to the error described [here](https://github.com/mafintosh/end-of-stream/issues/10).

I get this error:
```
Error: ENOENT: no such file or directory, open '/Users/cmccarthy/.nvm/versions/node/v6.10.0/lib/node_modules/yarn/node_modules/end-of-stream/node_modules/once/once.js'
```

This was working yesterday afternoon and it looks like `end-of-stream` pushed an update to 1.4.0 just 4 hours ago that is causing this.

**If the current behavior is a bug, please provide the steps to reproduce.**
In any node app:
```
npm install -g yarn
yarn install
```

**What is the expected behavior?**
yarn install works

I think yarn could pin `end-of-stream` to 1.2.0 to fix it.

**Please mention your node.js, yarn and operating system version.**
node 6.10.0 and npm 3.10.10

**Comment by OndraZizka** [comment]
Just looking at the same issue like "huh?"

**Comment by code0100fun** [comment]
It's a hack but if anyone is stuck with a failing CI due to this you can do this to get around it for now:

```
export BACK_TO_WORKSPACE=$(pwd)
cd $(npm config get prefix)/lib/node_modules/yarn/node_modules/end-of-stream
npm install once
cd $BACK_TO_WORKSPACE
```

**Comment by perplexes** [comment]
We're seeing this on Travis as of yesterday.

**Comment by luizfilipe** [comment]
same problem here

**Comment by jsight** [comment]
Floating versions and non-deterministic builds appear to have broken the internet again. I still wonder how this happens with yarn, though.

**Comment by mexanoid** [comment]
Same issue. Hack did not help.

**Comment by BradfordMedeiros** [comment]
:(

**Comment by AlexLandau** [comment]
I'm seeing this when yarn is installed via npm. Has anyone seen this error when other methods of installation are used? (To @jsight's comment, the problem is that we can't install yarn with yarn itself.)

**Comment by jasonmit** [comment]
@jsight yarn discourages installing via npm since they cannot guarantee the dep tree.

**Comment by mexanoid** [comment]
Actually the issue cause by end-of-stream dependency [https://github.com/mafintosh/end-of-stream/releases](url). It was updated 13 hours ago.

**Comment by jsight** [comment]
@AlexLandau - Ah, good point. I guess that is our problem as well. Maybe the tool could emphasize this somehow?

**Comment by howardjing** [comment]
We ran into this issue as well -- we worked around it by specifying a yarn version when installing yarn via npm:

```
npm install -g yarn@v0.22.0
yarn install
```

**Comment by jsight** [comment]
Installing yarn properly fixed it here (rpm install instead of via npm).

**Comment by zman2245** [comment]
"Installing yarn properly fixed it here (rpm install instead of via npm)."

I'm running into this using the Gradle node plugin. Not sure, but seems that might be using npm. Any suggestions for working around that?

**Comment by jsight** [comment]
@zman2245 - There is a tarball available here: https://yarnpkg.com/en/docs/install#alternatives-tab

Perhaps Gradle could be configured to use it?

**Comment by jroitgrund** [comment]
Why not use an `npm-shrinkwrap.json` file?

**Comment by doublesharp** [comment]
I was able to get my Jenkins builds working again by specifying `yarn@0.22.0` in the global package install instead of just `yarn`.

<img width="801" alt="screen shot 2017-03-16 at 10 25 44 am" src="https://cloud.githubusercontent.com/assets/571472/24009733/1c584728-0a33-11e7-8257-2e93605b7213.png">

**Comment by llwt** [comment]
@doublesharp Can confirm that specifying @0.22.0 fixes our CI builds.

**Comment by jonathanong** [comment]
@llwt same here

**Comment by code0100fun** [comment]
I can confirm that 0.22.0 fixes my issue.

**Comment by zman2245** [comment]
Working so far for my CI environment as well.

**Comment by ppoisson** [comment]
This was the fix for me too...

```
npm install -g yarn@v0.22.0
yarn install
```
Thanks!

**Comment by marsbergen** [comment]
I can confirm that the fix with forcing to 0.22.0 works on our Jenkins server too

**Comment by codler** [comment]
I hope 0.22.0 will be released soon

**Comment by DanielJoyce** [comment]
Can this stop breaking please? So annoying. Gradle and Ant installs never break for me.

**Comment by jasonmit** [comment]
The docs explicitly call out that installing yarn via npm cannot be guaranteed. That's the very reason why yarn existed.

Please lock this thread.

**Comment by jsight** [comment]
@jasonmit - +1 - although I wonder if there could be a warning when it is installed this way? It seems to be a common cause of spurious issue reports.

**Comment by DanielJoyce** [comment]
Then remove yarn from npm if it's not supposed to be installed that way.

**Comment by jCube1980** [comment]
It is not an installation issue. I installed Yarn v0.21.3 using the windows installer and started encountering this issue after I did a "yarn upgrade".

**Comment by ffxsam** [comment]
What.. the... heck.

Yarn on Mac OS, installed via the official instructions (`brew install yarn`). Version 0.21.3, totally failing on `yarn init`, even `yarn --version`:

```
/usr/local/Cellar/yarn/0.21.3/libexec/lib/node_modules/yarn/bin/yarn.js:61
      throw err;
      ^

Error: ENOENT: no such file or directory, open '/usr/local/Cellar/yarn/0.21.3/libexec/lib/node_modules/yarn/node_modules/end-of-stream/node_modules/once/once.js'
    at Error (native)
    at Object.fs.openSync (fs.js:641:18)
    at Object.fs.readFileSync (fs.js:509:33)
    at Object.Module._extensions..js (module.js:578:20)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/usr/local/Cellar/yarn/0.21.3/libexec/lib/node_modules/yarn/node_modules/end-of-stream/index.js:1:74)
```

**Comment by SimenB** [comment]
I'm also seeing this on yarn from brew. The odd thing is that it worked, then I uninstalled it (I needed there to not be a global yarn to fix a bug in our build toolchain), and when I installed it, it stopped working.

Same tarball, since brew checks the sha, must the the formula itself.

EDIT: Reseting to this commit and reinstalling solves it: https://github.com/Homebrew/homebrew-core/blob/f92d1c1bb5165f56c6fcf2c6d7e189f4815564bc/Formula/yarn.rb

**Comment by Daniel15** [comment]
If anyone sees this again, can you please try deleting the `.roadrunner.json` from the Yarn cache directory? I suspect Roadrunner is doing something strange.

**Comment by SimenB** [comment]
I think the issue I'm seeing might be with the included npm with the version of node on brew (`7.7.4` yesterday, `7.8.0` now), and not so much a specific dependency anymore, since it actually install the distributed tarball. I've sent a PR to homebrew dropping the usage of npm install to install

**Comment by josheche** [comment]
In my experience, the best way to get around this is just run `rm '/usr/local/bin/node'` then re install node from nodejs.org. It always does the trick for me when I accidentally start messing with yarn in a project I'm using npm on.

- So `rm '/usr/local/bin/node'` removes nodejs from your local environment. 
- Re-installing it on their website is an order of magnitude easier than letting homebrew do a suplex on your symlinks and then piledrive you into dependency hell.
